// ngx-material-timepicker/esm2015/src/app/material-timepicker/components/timepicker-field/ngx-timepicker-field.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef, Input, TemplateRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { NgxMaterialTimepickerService } from '../../services/ngx-material-timepicker.service';
import { Subject } from 'rxjs';
import { getHours, getMinutes } from '../../utils/timepicker-time.utils';
import { TimeUnit } from '../../models/time-unit.enum';
import { takeUntil } from 'rxjs/operators';
import { TimeFormatterPipe } from '../../pipes/time-formatter.pipe';
export class NgxTimepickerFieldComponent {
    /**
     * @param {?} timepickerService
     */
    constructor(timepickerService) {
        this.timepickerService = timepickerService;
        this.minHour = 1;
        this.maxHour = 12;
        this.timeUnit = TimeUnit;
        this.buttonAlign = 'right';
        this._format = 12;
        this.unsubscribe$ = new Subject();
        this.onChange = (/**
         * @return {?}
         */
        () => {
        });
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set format(value) {
        this._format = value === 24 ? 24 : 12;
        this.minHour = this._format === 12 ? 1 : 0;
        this.maxHour = this._format === 12 ? 12 : 23;
    }
    /**
     * @return {?}
     */
    get format() {
        return this._format;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set defaultTime(val) {
        this.timepickerService.setDefaultTimeIfAvailable(val, null, null, this._format);
        this._defaultTime = val;
        this.timepickerTime = val;
    }
    /**
     * @return {?}
     */
    get defaultTime() {
        return this._defaultTime;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.period$ = this.timepickerService.selectedPeriod;
        this.timepickerService.selectedHour.pipe(takeUntil(this.unsubscribe$))
            .subscribe((/**
         * @param {?} hour
         * @return {?}
         */
        hour => this.hour = hour.time));
        this.timepickerService.selectedMinute.pipe(takeUntil(this.unsubscribe$))
            .subscribe((/**
         * @param {?} minute
         * @return {?}
         */
        minute => this.minute = minute.time));
        this.hoursList = getHours(this._format);
        this.minutesList = getMinutes();
    }
    /**
     * @param {?} val
     * @return {?}
     */
    writeValue(val) {
        if (val) {
            this.defaultTime = val;
        }
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.onChange = fn;
    }
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    setDisabledState(isDisabled) {
        this.disabled = isDisabled;
    }
    /**
     * @param {?} hour
     * @return {?}
     */
    changeHour(hour) {
        this.timepickerService.hour = this.hoursList.find((/**
         * @param {?} h
         * @return {?}
         */
        h => h.time === hour));
        this.changeTime();
    }
    /**
     * @param {?} minute
     * @return {?}
     */
    changeMinute(minute) {
        this.timepickerService.minute = this.minutesList.find((/**
         * @param {?} m
         * @return {?}
         */
        m => m.time === minute));
        this.changeTime();
    }
    /**
     * @param {?} period
     * @return {?}
     */
    changePeriod(period) {
        this.timepickerService.period = period;
        this.changeTime();
    }
    /**
     * @param {?} time
     * @return {?}
     */
    onTimeSet(time) {
        this.defaultTime = time;
        this.onChange(time);
        this.formatTime();
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.unsubscribe$.next();
    }
    /**
     * @private
     * @return {?}
     */
    changeTime() {
        /** @type {?} */
        const time = this.timepickerService.getFullTime(this._format);
        this.timepickerTime = time;
        this.onChange(time);
    }
    /**
     * @private
     * @return {?}
     */
    formatTime() {
        this.hour = new TimeFormatterPipe().transform(this.hour, TimeUnit.HOUR);
        this.minute = new TimeFormatterPipe().transform(this.minute, TimeUnit.MINUTE);
    }
}
NgxTimepickerFieldComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngx-timepicker-field',
                template: "<div class=\"ngx-timepicker\">\n    <ngx-timepicker-time-control\n        class=\"ngx-timepicker__control--first\"\n        [placeholder]=\"'HH'\"\n        [time]=\"hour\"\n        [min]=\"minHour\"\n        [max]=\"maxHour\"\n        [timeUnit]=\"timeUnit.HOUR\"\n        [disabled]=\"disabled\"\n        (timeChanged)=\"changeHour($event)\"></ngx-timepicker-time-control>\n    <span class=\"ngx-timepicker__time-colon ngx-timepicker__control--second\">:</span>\n    <ngx-timepicker-time-control\n        class=\"ngx-timepicker__control--third\"\n        [placeholder]=\"'MM'\"\n        [time]=\"minute\"\n        [min]=\"0\"\n        [max]=\"59\"\n        [timeUnit]=\"timeUnit.MINUTE\"\n        [disabled]=\"disabled\"\n        (timeChanged)=\"changeMinute($event)\"></ngx-timepicker-time-control>\n    <ngx-timepicker-period-selector\n        class=\"ngx-timepicker__control--forth\"\n        [selectedPeriod]=\"period$|async\"\n        [disabled]=\"disabled\"\n        (periodSelected)=\"changePeriod($event)\"\n        *ngIf=\"format !== 24\"></ngx-timepicker-period-selector>\n    <ngx-material-timepicker-toggle\n        class=\"ngx-timepicker__toggle\"\n        *ngIf=\"!controlOnly\"\n        [ngClass]=\"{'ngx-timepicker__toggle--left': buttonAlign === 'left'}\"\n        [for]=\"timepicker\"\n        [disabled]=\"disabled\">\n        <span ngxMaterialTimepickerToggleIcon>\n            <!--suppress HtmlUnknownAttribute -->\n            <ng-container *ngTemplateOutlet=\"toggleIcon || defaultIcon\"></ng-container>\n        </span>\n    </ngx-material-timepicker-toggle>\n</div>\n<ngx-material-timepicker\n    [ngxMaterialTimepickerTheme]=\"clockTheme\"\n    [defaultTime]=\"timepickerTime\"\n    [format]=\"format\"\n    (timeSet)=\"onTimeSet($event)\" #timepicker></ngx-material-timepicker>\n\n<ng-template #defaultIcon>\n    <svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 24 24\" width=\"24px\" height=\"24px\">\n        <path\n            d=\"M 12 2 C 6.4889971 2 2 6.4889971 2 12 C 2 17.511003                   6.4889971 22 12 22 C 17.511003 22 22 17.511003 22 12 C 22 6.4889971 17.511003 2 12 2 z M 12 4 C 16.430123 4 20 7.5698774 20 12 C 20 16.430123 16.430123 20 12 20 C 7.5698774 20 4 16.430123 4 12 C 4 7.5698774 7.5698774 4 12 4 z M 11 6 L 11 12.414062 L 15.292969 16.707031 L 16.707031 15.292969 L 13 11.585938 L 13 6 L 11 6 z\"/>\n    </svg>\n</ng-template>\n",
                providers: [
                    NgxMaterialTimepickerService,
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => NgxTimepickerFieldComponent)),
                        multi: true
                    }
                ],
                styles: [".ngx-timepicker{display:-webkit-box;display:flex;-webkit-box-align:center;align-items:center;height:100%;border-bottom:1px solid rgba(0,0,0,.12)}.ngx-timepicker__time-colon{margin-left:10px}.ngx-timepicker__control--first{-webkit-box-ordinal-group:2;order:1}.ngx-timepicker__control--second{-webkit-box-ordinal-group:3;order:2}.ngx-timepicker__control--third{-webkit-box-ordinal-group:4;order:3}.ngx-timepicker__control--forth,.ngx-timepicker__toggle{-webkit-box-ordinal-group:5;order:4}.ngx-timepicker__toggle--left{-webkit-box-ordinal-group:1;order:0}"]
            }] }
];
/** @nocollapse */
NgxTimepickerFieldComponent.ctorParameters = () => [
    { type: NgxMaterialTimepickerService }
];
NgxTimepickerFieldComponent.propDecorators = {
    disabled: [{ type: Input }],
    toggleIcon: [{ type: Input }],
    buttonAlign: [{ type: Input }],
    clockTheme: [{ type: Input }],
    controlOnly: [{ type: Input }],
    format: [{ type: Input }],
    defaultTime: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    NgxTimepickerFieldComponent.prototype.period$;
    /** @type {?} */
    NgxTimepickerFieldComponent.prototype.hour;
    /** @type {?} */
    NgxTimepickerFieldComponent.prototype.minute;
    /** @type {?} */
    NgxTimepickerFieldComponent.prototype.minHour;
    /** @type {?} */
    NgxTimepickerFieldComponent.prototype.maxHour;
    /** @type {?} */
    NgxTimepickerFieldComponent.prototype.timeUnit;
    /** @type {?} */
    NgxTimepickerFieldComponent.prototype.timepickerTime;
    /** @type {?} */
    NgxTimepickerFieldComponent.prototype.disabled;
    /** @type {?} */
    NgxTimepickerFieldComponent.prototype.toggleIcon;
    /** @type {?} */
    NgxTimepickerFieldComponent.prototype.buttonAlign;
    /** @type {?} */
    NgxTimepickerFieldComponent.prototype.clockTheme;
    /** @type {?} */
    NgxTimepickerFieldComponent.prototype.controlOnly;
    /**
     * @type {?}
     * @private
     */
    NgxTimepickerFieldComponent.prototype._defaultTime;
    /**
     * @type {?}
     * @private
     */
    NgxTimepickerFieldComponent.prototype._format;
    /**
     * @type {?}
     * @private
     */
    NgxTimepickerFieldComponent.prototype.hoursList;
    /**
     * @type {?}
     * @private
     */
    NgxTimepickerFieldComponent.prototype.minutesList;
    /**
     * @type {?}
     * @private
     */
    NgxTimepickerFieldComponent.prototype.unsubscribe$;
    /**
     * @type {?}
     * @private
     */
    NgxTimepickerFieldComponent.prototype.onChange;
    /**
     * @type {?}
     * @private
     */
    NgxTimepickerFieldComponent.prototype.timepickerService;
}
